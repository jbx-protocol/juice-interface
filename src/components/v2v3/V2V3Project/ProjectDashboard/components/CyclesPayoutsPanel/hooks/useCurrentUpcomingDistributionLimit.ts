import { useProjectContext } from 'components/v2v3/V2V3Project/ProjectDashboard/hooks/useProjectContext'
import { useProjectMetadataContext } from 'contexts/shared/ProjectMetadataContext'
import useProjectDistributionLimit from 'hooks/v2v3/contractReader/useProjectDistributionLimit'
import { useProjectUpcomingFundingCycle } from 'hooks/v2v3/contractReader/useProjectUpcomingFundingCycle'

export const useCurrentUpcomingDistributionLimit = (
  type: 'current' | 'upcoming',
) => {
  const { projectId } = useProjectMetadataContext()
  const {
    primaryETHTerminal,
    distributionLimit: currentDistributionLimit,
    distributionLimitCurrency: currentDistributionLimitCurrency,
  } = useProjectContext()

  const {
    data: upcomingFundingCycleData,
    loading: upcomingFundingCycleLoading,
  } = useProjectUpcomingFundingCycle({ projectId })
  const [upcomingFundingCycle] = upcomingFundingCycleData ?? []

  const {
    data: upcomingDistributionLimitData,
    loading: upcomingDistributionLimitLoading,
  } = useProjectDistributionLimit({
    projectId,
    configuration: upcomingFundingCycle?.configuration.toString(),
    terminal: primaryETHTerminal,
  })

  const [upcomingDistributionLimit, upcomingDistributionLimitCurrency] =
    upcomingDistributionLimitData ?? []

  if (type === 'current') {
    return {
      distributionLimit: currentDistributionLimit,
      distributionLimitCurrency: currentDistributionLimitCurrency,
      loading: false,
    }
  }
  return {
    distributionLimit: upcomingDistributionLimit,
    distributionLimitCurrency: upcomingDistributionLimitCurrency,
    loading: upcomingFundingCycleLoading || upcomingDistributionLimitLoading,
  }
}
