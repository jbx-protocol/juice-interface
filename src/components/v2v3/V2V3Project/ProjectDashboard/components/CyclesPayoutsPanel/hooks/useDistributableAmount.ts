import { V2V3ProjectContext } from 'contexts/v2v3/Project/V2V3ProjectContext'
import { V2V3CurrencyOption } from 'models/v2v3/currencyOption'
import { useContext } from 'react'
import { V2V3_CURRENCY_ETH } from 'utils/v2v3/currency'

export const useDistributableAmount = () => {
  const {
    distributionLimit,
    usedDistributionLimit,
    balanceInDistributionLimitCurrency,
    distributionLimitCurrency,
  } = useContext(V2V3ProjectContext)

  const currency = (
    distributionLimitCurrency
      ? Number(distributionLimitCurrency)
      : V2V3_CURRENCY_ETH
  ) as V2V3CurrencyOption

  const effectiveDistributionLimit = distributionLimit ?? BigInt(0)
  const distributedAmount = usedDistributionLimit ?? BigInt(0)
  const balInDistributionLimit = balanceInDistributionLimitCurrency ?? BigInt(0)

  const distributable =
    effectiveDistributionLimit === 0n
      ? effectiveDistributionLimit
      : effectiveDistributionLimit - distributedAmount

  const distributableAmount =
    balInDistributionLimit > distributable
      ? distributable
      : balInDistributionLimit

  return {
    distributableAmount,
    currency,
  }
}
