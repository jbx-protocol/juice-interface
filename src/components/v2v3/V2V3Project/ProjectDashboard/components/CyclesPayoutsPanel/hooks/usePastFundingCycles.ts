import {
  FundingCycle_OrderBy,
  OrderDirection,
  useFundingCyclesQuery,
} from 'generated/graphql'
import { client } from 'lib/apollo/client'

const DEFAULT_PAGE_SIZE = 5

export function usePastFundingCycles({
  projectId,
  pageSize,
}: {
  projectId: number | undefined
  pageSize?: number
}) {
  return useFundingCyclesQuery({
    client,
    variables: {
      where: {
        projectId,
        endTimestamp_lt: Math.floor(Date.now() / 1000), // Only already ended
      },
      orderBy: FundingCycle_OrderBy.number,
      orderDirection: OrderDirection.desc,
      first: pageSize ?? DEFAULT_PAGE_SIZE,
    },
  })
}
