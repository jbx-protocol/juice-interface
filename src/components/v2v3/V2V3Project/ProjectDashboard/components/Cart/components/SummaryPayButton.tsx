import { Trans } from '@lingui/macro'
import { Button } from 'antd'
import { useCartSummary } from '../hooks/useCartSummary'

export const SummaryPayButton = ({ className }: { className?: string }) => {
  const { payProject, walletConnected } = useCartSummary()
  return (
    <Button className={className} type="primary" onClick={payProject}>
      <span>
        {walletConnected ? (
          <Trans>Pay project</Trans>
        ) : (
          <Trans>Connect wallet</Trans>
        )}
      </span>
    </Button>
  )
}
