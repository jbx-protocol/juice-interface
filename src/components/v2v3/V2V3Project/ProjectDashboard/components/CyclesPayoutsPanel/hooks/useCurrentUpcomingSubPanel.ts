import { t } from '@lingui/macro'
import { useFundingCycleCountdown } from 'components/v2v3/V2V3Project/ProjectDashboard/hooks/useFundingCycleCountdown'
import { useProjectContext } from 'components/v2v3/V2V3Project/ProjectDashboard/hooks/useProjectContext'
import { timeSecondsToDateString } from 'components/v2v3/V2V3Project/ProjectDashboard/utils/timeSecondsToDateString'
import { useProjectMetadataContext } from 'contexts/shared/ProjectMetadataContext'
import { useProjectUpcomingFundingCycle } from 'hooks/v2v3/contractReader/useProjectUpcomingFundingCycle'
import { BallotState } from 'models/v2v3/fundingCycle'
import { useMemo } from 'react'

export const useCurrentUpcomingSubPanel = (type: 'current' | 'upcoming') => {
  const { projectId } = useProjectMetadataContext()
  const {
    fundingCycle,
    loading: { fundingCycleLoading },
  } = useProjectContext()
  const {
    data: upcomingFundingCycleData,
    loading: upcomingFundingCycleLoading,
  } = useProjectUpcomingFundingCycle({
    projectId,
    /**
     * if the current cycle is unlocked, force the use of latestConfiguredFundingCycleOf.
     */
    useLatestConfigured: fundingCycle?.duration?.isZero() ?? false,
  })
  const [upcomingFundingCycle, , ballotState] = upcomingFundingCycleData ?? []
  const { timeRemainingText } = useFundingCycleCountdown()

  const cycleNumber = useMemo(() => {
    if (type === 'current') {
      return fundingCycle?.number.toNumber()
    }
    return fundingCycle?.number ? fundingCycle.number.toNumber() + 1 : undefined
  }, [fundingCycle?.number, type])

  const cycleUnlocked = useMemo(() => {
    if (type === 'current') {
      return fundingCycle?.duration?.isZero() ?? true
    }
    return upcomingFundingCycle?.duration?.isZero() ?? true
  }, [fundingCycle?.duration, type, upcomingFundingCycle?.duration])

  const upcomingCycleLength = useMemo(() => {
    if (!upcomingFundingCycle) return
    if (cycleUnlocked) return '-'
    return timeSecondsToDateString(
      upcomingFundingCycle.duration.toNumber(),
      'short',
    )
  }, [cycleUnlocked, upcomingFundingCycle])

  /** Determines if the CURRENT cycle is unlocked.
   * This is used to check if the upcoming cycle can start at any time. */
  const currentCycleUnlocked = fundingCycle?.duration?.isZero() ?? true

  const status = cycleUnlocked ? t`Unlocked` : t`Locked`
  const remainingTime = cycleUnlocked ? '-' : timeRemainingText

  // Short circuit current for faster loading
  if (type === 'current') {
    if (fundingCycleLoading) return { loading: true, type }
    return {
      loading: false,
      type,
      cycleNumber,
      status,
      remainingTime,
    }
  }

  if (fundingCycleLoading || upcomingFundingCycleLoading)
    return {
      loading: true,
      type,
    }

  return {
    loading: false,
    type,
    cycleNumber,
    status,
    cycleLength: upcomingCycleLength,
    cycleUnlocked,
    currentCycleUnlocked,
    hasPendingConfiguration:
      /**
       * If a cycle is unlocked, it may have a pending change.
       * The only way it would, is if the ballot state of the latestConfiguredFC is `approved`.
       */
      cycleUnlocked &&
      typeof ballotState !== 'undefined' &&
      ballotState !== null &&
      ballotState === BallotState.approved,
  }
}
