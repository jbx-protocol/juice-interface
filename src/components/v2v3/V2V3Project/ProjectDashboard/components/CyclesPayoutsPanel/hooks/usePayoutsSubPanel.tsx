import { AmountInCurrency } from 'components/currency/AmountInCurrency'
import { useProjectContext } from 'components/v2v3/V2V3Project/ProjectDashboard/hooks/useProjectContext'
import { Split } from 'models/splits'
import { V2V3CurrencyOption } from 'models/v2v3/currencyOption'
import { useCallback, useMemo } from 'react'
import assert from 'utils/assert'
import { getProjectOwnerRemainderSplit } from 'utils/splits'
import { V2V3CurrencyName } from 'utils/v2v3/currency'
import { isJuiceboxProjectSplit } from 'utils/v2v3/distributions'
import { isInfiniteDistributionLimit } from 'utils/v2v3/fundingCycle'
import {
  SPLITS_TOTAL_PERCENT,
  feeForAmount,
  formatSplitPercent,
} from 'utils/v2v3/math'
import { useCurrentUpcomingDistributionLimit } from './useCurrentUpcomingDistributionLimit'
import { useCurrentUpcomingPayoutSplits } from './useCurrentUpcomingPayoutSplits'
import { useDistributableAmount } from './useDistributableAmount'

const splitHasFee = (split: Split) => {
  return !isJuiceboxProjectSplit(split)
}

const calculateSplitAmountWad = (
  split: Split,
  distributionLimit: bigint | undefined,
  primaryETHTerminalFee: bigint | undefined,
) => {
  const splitValue = distributionLimit
    ? (distributionLimit * BigInt(split.percent)) / SPLITS_TOTAL_PERCENT
    : undefined
  const feeAmount = splitHasFee(split)
    ? feeForAmount(splitValue, primaryETHTerminalFee) ?? BigInt(0)
    : BigInt(0)
  return splitValue ? splitValue - feeAmount : undefined
}

export const usePayoutsSubPanel = (type: 'current' | 'upcoming') => {
  const { splits, loading } = useCurrentUpcomingPayoutSplits(type)
  const { projectOwnerAddress, primaryETHTerminalFee } = useProjectContext()
  const { distributableAmount } = useDistributableAmount()

  const { distributionLimit, distributionLimitCurrency } =
    useCurrentUpcomingDistributionLimit(type)

  const showAmountOnPayout =
    distributionLimit === 0n || isInfiniteDistributionLimit(distributionLimit)
      ? false
      : true

  const transformSplit = useCallback(
    (split: Split) => {
      assert(split.beneficiary, 'Beneficiary must be defined')
      let amount = undefined
      const splitAmountWad = calculateSplitAmountWad(
        split,
        distributionLimit,
        primaryETHTerminalFee,
      )
      if (showAmountOnPayout && splitAmountWad && distributionLimitCurrency) {
        amount = (
          <AmountInCurrency
            amount={splitAmountWad}
            currency={V2V3CurrencyName(
              Number(distributionLimitCurrency) as V2V3CurrencyOption,
            )}
          />
        )
      }
      return {
        projectId: split.projectId
          ? Number(BigInt(split.projectId))
          : undefined,
        address: split.beneficiary!,
        percent: `${formatSplitPercent(BigInt(split.percent))}%`,
        amount,
      }
    },
    [
      distributionLimit,
      distributionLimitCurrency,
      primaryETHTerminalFee,
      showAmountOnPayout,
    ],
  )

  const totalPayoutAmount = useMemo(() => {
    if (!distributionLimit || !distributionLimitCurrency) return
    if (
      isInfiniteDistributionLimit(distributionLimit) ||
      distributionLimit === 0n
    )
      return

    return (
      <AmountInCurrency
        amount={distributionLimit}
        currency={V2V3CurrencyName(
          Number(distributionLimitCurrency) as V2V3CurrencyOption,
        )}
      />
    )
  }, [distributionLimit, distributionLimitCurrency])

  const payouts = useMemo(() => {
    if (loading || !splits) return

    if (
      // We don't need to worry about upcoming as this is informational only
      type === 'current' &&
      splits.length === 0 &&
      distributableAmount === 0n
    ) {
      return []
    }

    const ownerPayout = projectOwnerAddress
      ? getProjectOwnerRemainderSplit(projectOwnerAddress, splits)
      : undefined

    return [
      ...splits,
      ...(ownerPayout && ownerPayout.percent > 0 ? [ownerPayout] : []),
    ]
      .sort((a, b) => Number(b.percent) - Number(a.percent))
      .map(transformSplit)
  }, [
    distributableAmount,
    loading,
    projectOwnerAddress,
    splits,
    transformSplit,
    type,
  ])

  return {
    loading,
    payouts,
    totalPayoutAmount,
    distributionLimit,
  }
}
