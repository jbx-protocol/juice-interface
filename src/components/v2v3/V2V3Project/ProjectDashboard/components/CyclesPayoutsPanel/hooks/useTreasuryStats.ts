import { t } from '@lingui/macro'
import { useProjectContext } from 'components/v2v3/V2V3Project/ProjectDashboard/hooks'
import { useCurrencyConverter } from 'hooks/useCurrencyConverter'
import { V2V3CurrencyOption } from 'models/v2v3/currencyOption'
import { useMemo } from 'react'
import { fromWad } from 'utils/format/formatNumber'
import { formatCurrencyAmount } from 'utils/formatCurrencyAmount'
import { V2V3_CURRENCY_ETH } from 'utils/v2v3/currency'
import { isInfiniteDistributionLimit } from 'utils/v2v3/fundingCycle'
import { useDistributableAmount } from './useDistributableAmount'

export const useTreasuryStats = () => {
  const {
    distributionLimit,
    distributionLimitCurrency: distributionLimitCurrencyRaw,
    balanceInDistributionLimitCurrency,
    primaryTerminalCurrentOverflow,
    fundingCycleMetadata,
  } = useProjectContext()
  const { distributableAmount } = useDistributableAmount()

  const distributionLimitCurrency: V2V3CurrencyOption | undefined =
    useMemo(() => {
      if (!distributionLimitCurrencyRaw) return undefined
      if (distributionLimitCurrencyRaw.eq(0)) return V2V3_CURRENCY_ETH // treat as eth
      return distributionLimitCurrencyRaw.toNumber() as V2V3CurrencyOption
    }, [distributionLimitCurrencyRaw])

  const treasuryBalance = useMemo(() => {
    if (!balanceInDistributionLimitCurrency) return undefined
    return formatCurrencyAmount({
      amount: Number(fromWad(balanceInDistributionLimitCurrency)),
      currency: distributionLimitCurrency,
    })
  }, [balanceInDistributionLimitCurrency, distributionLimitCurrency])

  const { weiToUsd } = useCurrencyConverter()

  const overflow = useMemo(() => {
    if (!distributionLimit) return undefined
    if (isInfiniteDistributionLimit(distributionLimit)) return t`No overflow`

    const overflowInDistributionLimitCurrency =
      distributionLimitCurrency === V2V3_CURRENCY_ETH
        ? fromWad(primaryTerminalCurrentOverflow)
        : weiToUsd(primaryTerminalCurrentOverflow)

    return formatCurrencyAmount({
      amount: overflowInDistributionLimitCurrency?.toString(),
      currency: distributionLimitCurrency,
    })
  }, [
    primaryTerminalCurrentOverflow,
    distributionLimitCurrency,
    distributionLimit,
    weiToUsd,
  ])

  const availableToPayout = useMemo(() => {
    return formatCurrencyAmount({
      amount: Number(fromWad(distributableAmount)),
      currency: distributionLimitCurrency,
    })
  }, [distributableAmount, distributionLimitCurrency])
  return {
    treasuryBalance,
    availableToPayout,
    overflow,
    redemptionRate: fundingCycleMetadata?.redemptionRate,
  }
}
