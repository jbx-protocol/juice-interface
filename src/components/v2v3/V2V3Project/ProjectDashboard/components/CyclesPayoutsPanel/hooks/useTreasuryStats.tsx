import { t } from '@lingui/macro'
import { AmountInCurrency } from 'components/currency/AmountInCurrency'
import { useProjectContext } from 'components/v2v3/V2V3Project/ProjectDashboard/hooks'
import { useCurrencyConverter } from 'hooks/useCurrencyConverter'
import { V2V3CurrencyOption } from 'models/v2v3/currencyOption'
import { useMemo } from 'react'
import { parseWad } from 'utils/format/formatNumber'
import { V2V3CurrencyName, V2V3_CURRENCY_ETH } from 'utils/v2v3/currency'
import { isInfiniteDistributionLimit } from 'utils/v2v3/fundingCycle'
import { useDistributableAmount } from './useDistributableAmount'

export const useTreasuryStats = () => {
  const {
    distributionLimit,
    distributionLimitCurrency: distributionLimitCurrencyRaw,
    balanceInDistributionLimitCurrency,
    primaryTerminalCurrentOverflow,
    fundingCycleMetadata,
  } = useProjectContext()
  const { distributableAmount } = useDistributableAmount()

  const distributionLimitCurrency: V2V3CurrencyOption | undefined =
    useMemo(() => {
      if (!distributionLimitCurrencyRaw) return undefined
      if (distributionLimitCurrencyRaw.eq(0)) return V2V3_CURRENCY_ETH // treat as eth
      return distributionLimitCurrencyRaw.toNumber() as V2V3CurrencyOption
    }, [distributionLimitCurrencyRaw])

  const treasuryBalance = useMemo(() => {
    if (!balanceInDistributionLimitCurrency) return undefined

    return (
      <AmountInCurrency
        amount={balanceInDistributionLimitCurrency}
        currency={V2V3CurrencyName(distributionLimitCurrency)}
      />
    )
  }, [balanceInDistributionLimitCurrency, distributionLimitCurrency])

  const { weiToUsd } = useCurrencyConverter()

  const overflow = useMemo(() => {
    if (!distributionLimit) return undefined
    if (isInfiniteDistributionLimit(distributionLimit)) return t`No overflow`

    const overflowInDistributionLimitCurrency =
      distributionLimitCurrency === V2V3_CURRENCY_ETH
        ? primaryTerminalCurrentOverflow
        : parseWad(weiToUsd(primaryTerminalCurrentOverflow))

    return (
      <AmountInCurrency
        amount={overflowInDistributionLimitCurrency}
        currency={V2V3CurrencyName(distributionLimitCurrency)}
      />
    )
  }, [
    primaryTerminalCurrentOverflow,
    distributionLimitCurrency,
    distributionLimit,
    weiToUsd,
  ])

  const availableToPayout = useMemo(() => {
    return (
      <AmountInCurrency
        amount={distributableAmount}
        currency={V2V3CurrencyName(distributionLimitCurrency)}
      />
    )
  }, [distributableAmount, distributionLimitCurrency])
  return {
    treasuryBalance,
    availableToPayout,
    overflow,
    redemptionRate: fundingCycleMetadata?.redemptionRate,
  }
}
