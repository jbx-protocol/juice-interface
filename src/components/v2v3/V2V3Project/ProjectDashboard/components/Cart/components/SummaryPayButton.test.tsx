/**
 * @jest-environment jsdom
 */
import { render, screen } from '@testing-library/react'
import { useCartSummary } from '../hooks/useCartSummary'
import { SummaryPayButton } from './SummaryPayButton'

jest.mock('../hooks/useCartSummary')

describe('SummaryPayButton copy', () => {
  beforeEach(() => {
    ;(useCartSummary as jest.Mock).mockReturnValue({
      payProject: jest.fn(),
      walletConnected: true,
    })
  })

  it('should render', () => {
    const { container } = render(<SummaryPayButton />)
    expect(container).toMatchSnapshot()
  })

  it.each`
    walletConnected | buttonText
    ${true}         | ${'Pay project'}
    ${false}        | ${'Connect wallet'}
  `(
    'should render button with text $buttonText when walletConnected is $walletConnected',
    ({ walletConnected, buttonText }) => {
      ;(useCartSummary as jest.Mock).mockReturnValue({
        payProject: jest.fn(),
        walletConnected,
      })
      render(<SummaryPayButton />)
      expect(screen.getByRole('button')).toHaveTextContent(buttonText)
    },
  )
})
