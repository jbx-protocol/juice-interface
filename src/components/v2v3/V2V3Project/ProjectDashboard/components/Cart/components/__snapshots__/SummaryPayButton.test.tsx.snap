// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SummaryPayButton copy should render 1`] = `
<div>
  <button
    class="ant-btn ant-btn-primary"
    type="button"
  >
    <span>
      Pay project
    </span>
  </button>
</div>
`;
